[FIX] core: remove control characters from strings in xmlrpc

nda-odoo opened the pull request:
According to https://docs.python.org/3.7/library/xmlrpc.client.html

> When passing strings, characters special to XML such as <, >, and & will be automatically escaped. However, it’s the caller’s responsibility to ensure that the string is free of characters that aren’t allowed in XML, such as the control characters with ASCII values between 0 and 31 (except, of course, tab, newline and carriage return); failing to do this will result in an XML-RPC request that isn’t well-formed XML.

**steps to reproduce:**
- create a product with an ASCII control character in its name (ex: \x03)
- read the product name using XMLRPC

**before this commit:**
- client can't parse the response, an error is raised 
`xml.parsers.expat.ExpatError: not well-formed (invalid token)
`

**after this commit:**
- we make sure the string is free of those characters

opw-3617458

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

Julien00859 reviewed (changes requested):
The commit message doesn't explain why such control characters appeared in the first place, I had to read the ticket to understand that it is crap coming from an upgrade. Is it a recurring problem?

Cannot we cleanup the database of the customer to remove those obviously wrong control characters instead? I don't like the idea of running this regexp on every single char field that we are going to export via xmlrpc in the future.

nda-odoo commented:
Thanks for the detailed feedback @Julien00859 ! Before adapting the code to your recommendations I would like to make sure we're on the same page, because i'm not really sure what's the good solution here.

> Cannot we cleanup the database of the customer to remove those obviously wrong control characters instead?

Not really, it is (not only) an upgrade issue since you can easily reproduce it on a runbot, nothing in the framework (js or python) removes those characters if you uses them (and it works fine in the backend, you will only have an error during an xmlrpc read). One option is to remove these characters before saving the field to the database, but it's worth noting that jsonrpc doesn't pose this limitation so we might break some legitimate use case. And reading existing data would still get an error.

> I don't like the idea of running this regexp on every single char field that we are going to export via xmlrpc in the future.

I'm not a big fan of this either but i don't see how we could do it another way if we want to keep those characters for jsonrpc. We could also call it a limitation of xmlrpc and add a note to the documentation telling users to use jsonrpc instead for this use case? Because from a business point of view, blanking the string seems overkill. Maybe it's better to raise an error than silently sending an empty response.

bouvyd commented:
hi there

my 2cents:
1. xmlrpc calls are not that frequent, i see no issue with running the regexp on them to remove control characters
2. indeed sending blank info without any explanation is just weird

all in all i think it's worth doing the fix

i'm a bit intrigued by the difference in treatment between bytes and str here, but one can assume that bytes will usually represent binary data of some form i guess so it make sense not to clean them i guess

nda-odoo commented:
I took @bouvyd 2 cents and made it 2 dollars, so @Julien00859 If i understood correctly:
- we should apply the regex at every xmlrpc read
- we remove the character instead of blanking the field

I made the requested changes (see https://github.com/odoo/odoo/compare/a2da3df74ebdfe992884423aca1249402b00ddbf..b45b3a795d4d65443afe8efa33a05b321e21c26a ). I'm not sure if there is a cleaner way to handle the both compiled regex?

As for the commit message, I believe it's clear enough but maybe you wanted something more? Maybe the example used in the test?

Julien00859 commented:
> As for the commit message, I believe it's clear enough but maybe you wanted something more?

A short (one line) explanation of how those characters got there in the first place, otherwise the commit message is great :)

Otherwise I think it is ok this way, in case of strings we remove the wrong characters, since it should be readable text I guess it is ok. For bytes on the contrary, removing sometimes some characters can be wrong, like in case of a raw image, removing some characters can make the image un-readable so better to just blank the field.

I'll let @odoo/rd-framework-py have the final word :)

nda-odoo commented on odoo/addons/base/controllers/rpc.py:
```diff
@@ -33,7 +41,7 @@ def dump_bytes(self, value, write):
         # client-side parsers
-        if XML_INVALID.search(value):
+        if XML_INVALID_BINARY.search(value):
             self.dump_unicode('', write)
```
Oh ok then. Thanks!

nda-odoo commented on odoo/addons/base/controllers/rpc.py:
```diff
@@ -24,19 +23,21 @@ def dump_frozen_dict(self, value, write):
+            # client side parsers.
+            translation_table = {char: None for char in CONTROL_CHARACTERS}
+            return input_string.translate(translation_table)
+        return super().dump_unicode(remove_control_characters(value), write)
```
You're right it's cleaner that way.

Julien00859 reviewed (commented):
Would write a line inside the commit message to note that we changed dump_bytes for consistency, and cover that too with an unittest. I was a bit surprised that `str.translate` understood `{ord('a'): None}` as `{ord('a'): ''}` I would like to suggest `CONTROL_CHARACTERS = dict.fromkeys(set(range(32)) - {9, 10, 13}, value='')` but that's nitpicking.

nda-odoo commented:
Hello @Julien00859 

> Would write a line inside the commit message to note that we changed dump_bytes for consistency,

I updated the commit message to explain the change

> and cover that too with an unittest. 

I updated the existing one instead, to reflect the new behavior

Is something else missing?

Julien00859 reviewed (approved):
Thanks!

@robodoo r+
